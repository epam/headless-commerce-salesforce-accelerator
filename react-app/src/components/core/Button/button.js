/*
 * Copyright 2020 EPAM Systems, Inc. (https://www.epam.com/)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { forwardRef } from "react";
import PropTypes from "prop-types";

import { Link } from "react-router-dom";
import {
  StyledPrimaryButton,
  StyledLinkButton,
  StyledIcon,
  StyledSecondaryButton,
} from "./buttonStyled";

const BUTTON_VARIANTS = {
  primary: StyledPrimaryButton,
  link: StyledLinkButton,
  secondary: StyledSecondaryButton,
};

const Button = forwardRef(
  (
    {
      id,
      type = "text",
      disabled = false,
      icon = null,
      plain = false,
      onClick,
      onFocus,
      children,
      margin,
      size = "lg",
      link = "",
      variant = "primary",
      ...props
    },
    ref
  ) => {
    const StyledButton = BUTTON_VARIANTS[variant];

    return (
      <StyledButton
        id={id}
        type={type}
        ref={ref}
        disabled={disabled}
        onClick={onClick}
        onFocus={onFocus}
        isIconOnly={!children}
        margin={margin}
        plain={plain}
        size={size}
        link={!!link}
        {...props}
      >
        {link && <Link to={link} />}
        {children}
        {icon && <StyledIcon>{icon}</StyledIcon>}
      </StyledButton>
    );
  }
);

Button.propTypes = {
  id: PropTypes.number,
  type: PropTypes.string,
  size: PropTypes.string,
  link: PropTypes.string,
  style: PropTypes.string,
  variant: PropTypes.string,
  disabled: PropTypes.bool,
  plain: PropTypes.bool,
  onClick: PropTypes.func,
  onFocus: PropTypes.func,
  icon: PropTypes.element,
  children: PropTypes.node,
  margin: PropTypes.string,
  minWidth: PropTypes.string,
};

export default Button;
