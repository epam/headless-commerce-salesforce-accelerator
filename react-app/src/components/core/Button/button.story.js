/*
 * Copyright 2020 EPAM Systems, Inc. (https://www.epam.com/)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useEffect, useRef } from "react";
import { storiesOf } from "@storybook/react";
import { BrowserRouter as Router } from "react-router-dom";

import ThemeProvider from "../../../globalThemeProvider";
import Button from "./button";
import BagIcon from "../ShoppingBagIcon";

storiesOf("Button", module)
  .addDecorator((story) => (
    <ThemeProvider>
      <div style={{ margin: "20px", width: "900px" }}>{story()}</div>
    </ThemeProvider>
  ))
  .add("button regular large", () => <Button>Submit</Button>)
  .add("button regular small", () => <Button size="sm">Submit</Button>)
  .add("button icon only", () => <Button icon={<BagIcon />} />)
  .add("button plain icon only", () => <Button icon={<BagIcon />} plain />)
  .add("button iconed", () => <Button icon={<BagIcon />}>Add to Cart</Button>)
  .add("button iconed in focus", () =>
    React.createElement(() => {
      const ref = useRef(null);
      useEffect(() => {
        ref.current.focus();
      }, []);
      return (
        <Button ref={ref} icon={<BagIcon />}>
          Add to Cart
        </Button>
      );
    })
  )
  .add("button iconed disabled", () => (
    <Button disabled icon={<BagIcon />}>
      Add to Cart
    </Button>
  ))
  .add("button link", () => (
    <Router>
      <Button link="/home" variant="link">
        View Cart
      </Button>
    </Router>
  ))
  .add("button link in focus", () =>
    React.createElement(() => {
      const ref = useRef(null);
      useEffect(() => {
        ref.current.focus();
      }, []);
      return (
        <Router>
          <Button ref={ref} link="/home" variant="link">
            View Cart
          </Button>
        </Router>
      );
    })
  )

  .add("secondary button", () => (
    <Router>
      <Button variant="secondary">Cancel</Button>
    </Router>
  ))

  .add("button link disabled", () => (
    <Router>
      <Button link="/home" variant="link" disabled>
        View Cart
      </Button>
    </Router>
  ));
