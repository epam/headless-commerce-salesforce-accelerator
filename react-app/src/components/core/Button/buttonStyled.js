/*
 * Copyright 2020 EPAM Systems, Inc. (https://www.epam.com/)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import styled from "styled-components";

export const StyledIcon = styled.div`
  display: inline-block;
  margin-right: 8px;
  height: 24px;
`;

export const StyledPrimaryButton = styled.button`
  background-color: ${(props) =>
    props.plain ? "transparent" : props.theme.colors.wattle};
  color: ${(props) => props.theme.colors.mineShaft};
  font: ${(props) =>
    props.theme.font(
      "semiBold",
      (props.size === "sm" && 14) || (props.size === "lg" && 16),
      24
    )};
  margin: ${({ margin = 0 }) => margin};
  padding: ${(props) =>
    (props.plain && "0px") ||
    (props.isIconOnly && "8px") ||
    (props.size === "sm" && "8px 16px") ||
    (props.size === "lg" && "12px 32px")};
  border: none;
  display: flex;
  justify-content: center;
  min-width: ${({ minWidth }) => minWidth};
  cursor: pointer;
  border-radius: 2px;

  &:focus {
    outline: 1px solid ${(props) => props.theme.colors.semiWhite};
    outline-offset: -3px;
  }

  &:active {
    background-color: ${(props) => props.theme.colors.turmeric};
  }

  &:hover {
    background-color: ${(props) =>
      props.plain ? "transparent" : props.theme.colors.yellowGreen};
  }

  &:disabled {
    background-color: ${(props) => props.theme.colors.alabaster};
    border: 1px solid ${(props) => props.theme.colors.iron};
    color: ${(props) => props.theme.colors.dimGray};
    pointer-events: none;

    path {
      fill: ${(props) => props.theme.colors.dimGray};
    }
  }

  ${(props) =>
    props.isIconOnly &&
    `${StyledIcon} {
    margin-right: 0;
  }`}
`;

export const StyledLinkButton = styled.button`
  color: ${(props) => props.theme.colors.blue};
  font: ${(props) => props.theme.font("regular", 14, 24)};
  background-color: transparent;
  border: none;
  cursor: pointer;
  padding: 0 8px;
  height: 24px;
  display: flex;
  align-items: center;

  &:hover {
    text-decoration: underline;
  }

  &:focus {
    border: 1px solid ${(props) => props.theme.colors.blue};
    outline: none;
    border-radius: 2px;
  }

  &:disabled {
    color: ${(props) => props.theme.colors.silverChalice};
    pointer-events: none;
  }
`;

export const StyledSecondaryButton = styled(StyledPrimaryButton)`
  max-height: 48px;
  box-sizing: border-box;
  border: 1px solid ${({ theme }) => theme.colors.lightGray};
  background-color: ${({ theme }) => theme.colors.white};

  &:hover,
  &:active {
    background-color: ${({ theme }) => theme.colors.alabaster};
  }

  &:focus {
    outline-color: ${({ theme }) => theme.colors.wattle};
  }
`;
