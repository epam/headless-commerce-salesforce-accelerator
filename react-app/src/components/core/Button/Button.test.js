/*
 * Copyright 2020 EPAM Systems, Inc. (https://www.epam.com/)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react";
import { shallow } from "enzyme";
import Button from ".";
import getMainTheme from "../../../theme";

const defaultProps = {
  plain: true,
  isIconOnly: true,
};

const theme = getMainTheme();

describe("Button", () => {
  const wrapped = shallow(<Button />);
  it("should match snapshot", () => {
    expect(wrapped).toMatchSnapshot();
  });
});

describe("Button link", () => {
  const wrapped = shallow(
    <Button theme={theme} {...defaultProps} size="lg" variant="link" />
  );
  it("should match snapshot", () => {
    expect(wrapped.dive()).toMatchSnapshot();
  });
});

describe("Button secondary", () => {
  const wrapped = shallow(
    <Button theme={theme} {...defaultProps} size="lg" variant="secondary" />
  );
  it("should match snapshot", () => {
    expect(wrapped.dive()).toMatchSnapshot();
  });
});
