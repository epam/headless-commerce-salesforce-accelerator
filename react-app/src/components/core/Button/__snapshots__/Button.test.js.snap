// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match snapshot 1`] = `
<styled.button
  disabled={false}
  isIconOnly={true}
  link={false}
  plain={false}
  size="lg"
  type="text"
/>
`;
